### DISCUSSION TITLE
Repeat prepared statement usage causing high DB CPU usage

### DISCUSSION THREAD

**Discussion by mitchej123** [sqlalchemy/sqlalchemy#8384] [Usage Questions]
I'm still trying to nail down a minimum reproducible example but I haven't yet succeeded in finding anything except my production code that's tripping this issue.  I'm including an example that approximates the trigger.  

## Issue

I've got a large list of events that are being processed in batch that I need to check against the database for idempotency or upversioning.  One of the batches being processed is about 200k events, and coupled with the 32k limit on parameters for asyncpg, this means I need to batch the queries.  I'm doing around 22 select queries each with about 20k parameters `id IN (ids) or external_id IN (external_ids)`.  Queries 1-5 run fairly fast, and then starting with the 6th query the DB CPU usage spikes and the queries go from ~500ms to 3-5+ minutes.

I tried running the queries concurrently (on different connections) and they all finished quickly.  I tried running sequentially, on different connections, and they all finished quickly.  However when running sequentially on a single connection query 6 (and onwards) always slow way down.  I noticed that if I disable the prepared statement cache, the issue completely goes away.    My current workaround is on every 6th query, removing the prepared statement from the internal sqlalchemy cache, which is obviously less than ideal.    I haven't seen anything particularly interesting in the DB or application logs to suggest why this is happening.  

## Question

I'm curious if anyone else has run into this issue with prepared statements before, or suggestions on a better work around (such as perhaps being able to skip the prepared statement cache for specific queries, or the ability to invalidate the cached statement for a specific query).  The bug doesn't seem to be in SQLAlchemy or AsyncPG, but definitely seems to be triggered by their usage of prepared statements.

## Versions

Database Version: Google Cloud SQL - Postgres 14.4
SQLAlchemy Version: 1.4.40
AsyncPG Version: 0.26.0

## Approximate Example

```python
from __future__ import annotations

import asyncio
import itertools
import math
from uuid import uuid4

from sqlalchemy import Index, Integer, and_, func, or_
from sqlalchemy.dialects.postgresql import insert
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.future import select
from sqlalchemy.orm import sessionmaker
from structlog import get_logger

log = get_logger(__name__)

import sqlalchemy.dialects.postgresql
from sqlalchemy import Column, String
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()


class BigTable(Base):
    __tablename__ = "big_table"
    __table_args__ = (Index("ix_company_external_id", "company_id", "external_id"),)
    id = Column(sqlalchemy.dialects.postgresql.UUID(as_uuid=True), primary_key=True)
    company_id = Column(sqlalchemy.dialects.postgresql.UUID(as_uuid=True))
    external_id = Column(String(64))
    version = Column(Integer)


HOST = "localhost"
USER = "scott"
PASSWORD = "tiger"
DATABASE = "test"

MAX_PARAMS = 31500
DESIRED_COUNT = 3000000
OBJECTS_PER_INSERT = int(math.floor(MAX_PARAMS / 3))


def grouper(n: int, iterable):
    iterable = iter(iterable)
    return iter(lambda: list(itertools.islice(iterable, n)), [])


async def test_idempotency() -> None:
    engine = create_async_engine(f"postgresql+asyncpg://{USER}:{PASSWORD}@{HOST}/{DATABASE}", pool_size=300)
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)

    async_session = sessionmaker(engine, expire_on_commit=False, class_=AsyncSession)
    company_id = uuid4()

    async with async_session() as session:
        count = (await session.execute(select(func.count(BigTable.id)))).scalar_one()

    async def do_insert(i, num_inserts):
        async with async_session() as session:
            async with session.begin():
                print(f"Inserting batch {i}/{num_inserts}")
                query = insert(BigTable).values(
                    [{"id": uuid4(), "company_id": company_id, "external_id": str(uuid4())} for _ in range(OBJECTS_PER_INSERT)]
                )
                await session.execute(query)
        print(f"Done inserting batch {i}")

    if count < DESIRED_COUNT:
        num_inserts = int(math.ceil((DESIRED_COUNT - count) / OBJECTS_PER_INSERT))
        tasks = []
        for i in range(num_inserts):
            tasks.append(do_insert(i + 1, num_inserts))

        for my_tasks in grouper(10, tasks):
            # The inserts here are part of the setup to get a reasonable sized table; they are not part of the problem
            await asyncio.gather(*my_tasks)

    async with async_session() as session:
        print("Querying 10k objects to simulate a large overlapping batch being processed")
        items_to_query = (await session.execute(select(BigTable).limit(10000))).scalars().all()
        ids = set()
        external_ids = set()
        for item in items_to_query:
            ids.add(item.id)
            external_ids.add(item.external_id)

        query = (
            select(BigTable)
            .where(and_(BigTable.company_id == company_id, or_(BigTable.id.in_(ids), BigTable.external_id.in_(external_ids))))
            .order_by(BigTable.id)
        )

        for i in range(10):
            print(f"Querying batch {i}")
            # The 6th query (i=5) consistently spikes the CPU on the database and slows way down, despite the query
            # explaining fine
            await session.execute(query)


if __name__ == "__main__":
    loop = asyncio.get_event_loop()
    loop.run_until_complete(test_idempotency())

```

**Comment by zzzeek** [comment]
im not fully following where the problem is, if it's in the INSERT or the SELECT part.  just so I'm reading the code correctly, the step that runs `await asyncio.gather(*my_tasks)` runs all the INSERTs, and then that part is done.  then, it moves onto running some SELECT queries which are run by themselves, and it's the SELECT query you are having the problem with, is that correct?    When you said "One of the batches being processed is about 200k events, and coupled with the 32k limit on parameters for asyncpg, this means I need to batch the queries. I'm doing around 22 queries each with about 20k parameters. ", that sounds like you are referring to the INSERT part of it.  or are you referring to the parameters in the IN expression in your SELECT?

In the code as well I notice the rows aren't being fetched from asyncpg when the SELECT is run, I wouldn't expect that to cause an issue, but this does seem to be something related to how asyncpg's notion of "prepared statements" works.    We basically call [connection.prepare()](https://magicstack.github.io/asyncpg/current/api/index.html#asyncpg.connection.Connection.prepare) and with the cache we just stick that PreparedStatement object in a cache and use it again.  if you turn the cache off, then we don't do that.     now if there's some issue where we are not supposed to use the same object more than five times, something like that, I'm not really sure, but that's what you seem to be describing.   

the prepared statement cache works off the text of the SQL so if that were to change in some way, it would make a new prepared statement.  I dont have a convenient way to get arbitrary comment text into the SQL for you, it seems simpler just to turn off the prepared statement cache.   we can certainly have options to change its behavior on a per statement basis but the thing you're describing sounds pretty weird to start with.

**Reply by zzzeek** [reply]
I think also it's pretty important that when the program is hanging, you take a look inside the database pg_stat_activity view and see whats going on.  If the DB itself is where the CPU spike is, go look over there in the DB and Postgresql should be able to give you a pretty good view of what it's having trouble with.   also can this be reprodcued on other Postgresql databases not hosted by Google?  maybe some weird resource limit or bug on their end.

**Reply by mitchej123** [reply]
I included the INSERTS in the example trying to get a big enough table for the example;  The problem is triggering on query 6 out of 22 of the large SELECTS (with the IN clause).  

I didn't bother to read the query results in the sample code because it seems to be a database bug -- I use streaming in my code and it didn't start to send any results for a long time (the problem persisted with or without the streaming).   Once the query finally got to a place that it would start sending data, it went pretty quickly.

I did connect to the DB and looked at pg_stat_activity, but didn't see anything particularly exciting.  The queries were "active" the entire time, and the query start time/state change didn't move for about 10 minutes until it finally moved to the next one. I've reproduced it on two different cloud sql postgres instances; I'll look into reproducing it on a non google one.

> now if there's some issue where we are not supposed to use the same object more than five times

We've got a lot of other places that would definitely use the same cached query more than 5 times and aren't demonstrating this problem.  I haven't been able to narrow down why this particular query/table is different...  The sample code I wrote here is using the same query 10+ times and somehow not triggering the issue :exploding_head: 

> the prepared statement cache works off the text of the SQL so if that were to change in some way, it would make a new prepared statement. I dont have a convenient way to get arbitrary comment text into the SQL for you, it seems simpler just to turn off the prepared statement cache.

I had considered fuzzing the number of parameters, to accomplish this.


> we can certainly have options to change its behavior on a per statement basis but the thing you're describing sounds pretty weird to start with.

I agree, it's very weird.  I spent some time trying different versions of the query thinking it explained well with 100 params and just blew up with 20k, narrowing it down to the prepared statement usage was quite perplexing.  I still don't understand why it's repeatedly and consistently throwing a fit for this particular query on this particular table...

**Reply by mitchej123** [reply]
I did turn on query logging on the server for a while, and didn't see anything out of the ordinary.  It was doing `execute __asyncpg_stmt_e__/__asyncpg_portal__<#> <select statement here>` for the select queries, both the ones that spiked the CPU and the earlier ones that didn't.

**Reply by CaselIT** [reply]
Just to confirm, can you reporduce the issue with the example above, or it just illustrates what you are doing but it doesn't cause the bug?

**Reply by mitchej123** [reply]
The example above illustrates what I'm doing but does not yet trigger the bug.

I'm still trying to isolate exactly why.  My current hunch is the table/data because the code is almost identical.

**Reply by zzzeek** [reply]
can you make a new table and copy the data into it and see if *that* table has the problem?   then if that reproduces, try different variants of the table, like with less rows in it to start, different indexes, etc.   like if this isnt some internal state w/ the table.  also all the usual things like VACUUM etc., i dont know where PG is at these days with that stuff.

also reproducing on a different DB server not hosted by Google I think is an important thing to look into here as well.

**Comment by 20-buck-spin** [comment]
@mitchej123 

Were you able to fix this issue? I am currently facing a similar problem. Running the same query more than 5 times slows the query execution speed by almost half. I am not sure what is causing this problem and how to even tackle this. If i jump into psql and run the query, it's speed is similar to the first few times I ran the query through sqlalchemy

**Reply by CaselIT** [reply]
You would start by enabling logging using `echo=True` in the create_engine and see if the same query is executed like expected.
If so I'm not really sure what sqlalchemy can do here...

**Reply by 20-buck-spin** [reply]
Yes, the query is the same. I don't think this is an sqlalchemy issue. I just asked OP to comment because he had the same issue as me (the query being slower after the 5th query doesn't feel like a coincidence). I am looking into the driver (asynpg in my case) and maybe postgres settings hoping something there will help. Not sure what though because its the same query being run each time. I'll post an update to this thread once I find the cause (for anyone else who runs into this same issue)

**Reply by CaselIT** [reply]
My suggestion was just to verify that's indeed the same query and nothing strange was happening in the sqlalchemy side.

**Reply by 20-buck-spin** [reply]
I found the solution. This is a postgres settings issue (IDK about other DBs). Found the reason on https://stackoverflow.com/questions/77188652/query-become-slow-after-running-it-multiple-times-with-the-same-connection

TLDR you need to set force_custom_plan to force_custom_plan otherwise postgres changes the query execution plan on the fly automatically (the explain analyze output changes). My queries are running at the same speed now.
